core/inject.js

```diff
@@ -232,71 +232,163 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
+      }
+
+      return workspaceMetrics;
+    };
+    var getWorkspaceObjectMetrics = function(object) {
+      var position = object.getRelativeToSurfaceXY();
+      var size = object.getHeightWidth();
+
+      return {
+        // In LTR the position is at the top-left of the block, and in RTL
+        // it is at the top-right of the block. We need to add the width
+        // accordingly.
+        left: position.x - (options.RTL ? size.width : 0),
+        right: position.x + (options.RTL ? 0 : size.width),
+        top: position.y,
+        bottom: position.y + size.height,
+      };
+    };
+
+    var bumpObjects = function(e) {
+      // We always check isMovable_ again because the original
+      // "not movable" state of isMovable_ could have been changed.
+      if (!mainWorkspace.isDragging() && !mainWorkspace.isMovable() &&
+          (e.type == Blockly.Events.BLOCK_CREATE ||
+          e.type == Blockly.Events.BLOCK_MOVE ||
+          e.type == Blockly.Events.COMMENT_CREATE ||
+          e.type == Blockly.Events.COMMENT_MOVE)) {
+        var metrics = getWorkspaceMetrics();
+        if (metrics.contentTop < metrics.viewTop ||
+            metrics.contentBottom > metrics.viewBottom ||
+            metrics.contentLeft < metrics.viewLeft ||
+            metrics.contentRight > metrics.viewRight) {
+
+          // Handle undo.
           var oldGroup = null;
           if (e) {
             oldGroup = Blockly.Events.getGroup();
             Blockly.Events.setGroup(e.group);
           }
-          var movedBlocks = false;
-          for (var b = 0, block; block = blocks[b]; b++) {
-            var blockXY = block.getRelativeToSurfaceXY();
-            var blockHW = block.getHeightWidth();
-            // Bump any block that's above the top back inside.
-            var overflowTop = edgeTop + MARGIN - blockHW.height - blockXY.y;
-            if (overflowTop > 0) {
-              block.moveBy(0, overflowTop);
-              movedBlocks = true;
-            }
-            // Bump any block that's below the bottom back inside.
-            var overflowBottom =
-                edgeTop + metrics.viewHeight - MARGIN - blockXY.y;
-            if (overflowBottom < 0) {
-              block.moveBy(0, overflowBottom);
-              movedBlocks = true;
-            }
-            // Bump any block that's off the left back inside.
-            var overflowLeft = MARGIN + edgeLeft -
-                blockXY.x - (options.RTL ? 0 : blockHW.width);
-            if (overflowLeft > 0) {
-              block.moveBy(overflowLeft, 0);
-              movedBlocks = true;
-            }
-            // Bump any block that's off the right back inside.
-            var overflowRight = edgeLeft + metrics.viewWidth - MARGIN -
-                blockXY.x + (options.RTL ? blockHW.width : 0);
-            if (overflowRight < 0) {
-              block.moveBy(overflowRight, 0);
-              movedBlocks = true;
-            }
+
+          switch (e.type) {
```

Comment:
Should a getObjectById() function be added to the workspace so that we can remove this switch statement? Or do you think that's unnecessary?

Reply:
I think that's unnecessary--blocks and comments share a few APIs, but not many, so you'd still usually have to check the type of the object that you got back.